File: code_julia/utils/ComputeData.jl
#=============================================================================
Author: Rayla Kurosaki
GitHub: https://github.com/rkp1503
=============================================================================#

module ComputeData

import DifferentialEquations: ODEProblem, solve, Rodas5P
import ModelingToolkit: @named, ODESystem, structural_simplify
import OrderedCollections: OrderedDict

const ROUND = 3
const SOL_MAX = 10
const TOL_SOLUTION = 1e-12
    
function solve_model(model, D, variables_dict, parameters_dict, t, tₘₐₓ)
    vars_keys = collect(keys(variables_dict))
    vars_vals = collect(values(variables_dict))
    params_keys = collect(keys(parameters_dict))
    params_vals = collect(values(parameters_dict))
    @named sys = ODESystem(model(D, vars_keys, params_keys), t, vars_keys, params_keys)
    sys = structural_simplify(sys)
    prob = ODEProblem(sys, vars_vals, (0.0, tₘₐₓ), params_vals, jac = true)
    return solve(prob, Rodas5P(), reltol=TOL_SOLUTION, abstol=TOL_SOLUTION; verbose=false)
end;

function get_bifurcation_data(model, D, variables_dict, parameters, t, tₘₐₓ, bifurcation_parameter, parameter_bounds, parameter_values; data_points=10_000, debug=false)
	parameters_dict = OrderedDict(zip(parameters, parameter_values))
	x_arr_t, x_arr_s, x_arr_u = [], [], []
	y_arr_t, y_arr_s, y_arr_u = [], [], []
	z_arr_t, z_arr_s, z_arr_u = [], [], []
	prev_x, prev_y, prev_z = "", "", ""
	curr_x, curr_y, curr_z = "", "", ""
    bifurcation_values = range(parameter_bounds[1], parameter_bounds[2], length=1+data_points)
	for bifurcation_value in bifurcation_values
		parameters_dict[bifurcation_parameter] = bifurcation_value
		solutions = solve_model(model, D, variables_dict, parameters_dict, t, tₘₐₓ)
		n_start = floor(Int, length(solutions.t)*0.95)
		x_min, x_max = extrema(solutions'[:, 1][n_start:end])
		y_min, y_max = extrema(solutions'[:, 2][n_start:end])
		z_min, z_max = extrema(solutions'[:, 3][n_start:end])
		if ((1/data_points <= x_min < SOL_MAX) && (1/data_points <= x_max < SOL_MAX))
			push!(x_arr_t, [bifurcation_value, x_min])
			push!(x_arr_t, [bifurcation_value, x_max])
			if (abs(x_min-x_max) <= 1/data_points)
				curr_x = "Stable"
                push!(x_arr_s, [bifurcation_value, x_min])
                push!(x_arr_s, [bifurcation_value, x_max])
			else
				curr_x = "Unstable"
                push!(x_arr_u, [bifurcation_value, x_min])
                push!(x_arr_u, [bifurcation_value, x_max])
			end
		end
		if ((1/data_points <= y_min < SOL_MAX) && (1/data_points <= y_max < SOL_MAX))
			push!(y_arr_t, [bifurcation_value, y_min])
			push!(y_arr_t, [bifurcation_value, y_max])
			if (abs(y_min-y_max) <= 1/data_points)
				curr_y = "Stable"
                push!(y_arr_s, [bifurcation_value, y_min])
                push!(y_arr_s, [bifurcation_value, y_max])
			else
				curr_y = "Unstable"
                push!(y_arr_u, [bifurcation_value, y_min])
                push!(y_arr_u, [bifurcation_value, y_max])
			end
		end
		if ((1/data_points <= z_min < SOL_MAX) && (1/data_points <= z_max < SOL_MAX))
			push!(z_arr_t, [bifurcation_value, z_min])
			push!(z_arr_t, [bifurcation_value, z_max])
			if (abs(z_min-z_max) <= 1/data_points)
				curr_z = "Stable"
                push!(z_arr_s, [bifurcation_value, z_min])
                push!(z_arr_s, [bifurcation_value, z_max])
			else
				curr_z = "Unstable"
                push!(z_arr_u, [bifurcation_value, z_min])
                push!(z_arr_u, [bifurcation_value, z_max])
			end
		end
		if (curr_x == curr_y == curr_z) && (curr_x != prev_x)
			if debug
				if prev_x == ""
					println("Start -> $(curr_x): $(bifurcation_value)")
				else
					println("$(prev_x) -> $(curr_x): $(bifurcation_value)")
				end
			end
			prev_x = curr_x
			prev_y = curr_y
			prev_z = curr_z
		end
	end
    x_arr_t = reformat_data(x_arr_t)
    x_arr_s = reformat_data(x_arr_s)
    x_arr_u = reformat_data(x_arr_u)
	y_arr_t = reformat_data(y_arr_t)
    y_arr_s = reformat_data(y_arr_s)
    y_arr_u = reformat_data(y_arr_u)
	z_arr_t = reformat_data(z_arr_t)
    z_arr_s = reformat_data(z_arr_s)
    z_arr_u = reformat_data(z_arr_u)
    return (x_arr_t, x_arr_s, x_arr_u, y_arr_t, y_arr_s, y_arr_u, z_arr_t, z_arr_s, z_arr_u)
end;

function reformat_data(data)
    if data == []
		return []
	else
		return permutedims(reshape(hcat(data...), (length(data[1]), length(data))))
	end
end;

function approximate_bounds(model, D, variables_dict, parameters_dict, t, tₘₐₓ, equilibrium_type)
    solutions_x, solutions_y, solutions_z = [], [], []
    for _ in 1:MAX_ITER
        parameter_vals = generate_parameters_helper(parameters_dict, equilibrium_type; debug)
        parameters_dict_temp = OrderedDict(zip(keys(parameters_dict), parameter_vals))
        solutions = solve_model(model, D, variables_dict, parameters_dict_temp, t, tₘₐₓ)
        xₛₒₗ, yₛₒₗ, zₛₒₗ = last(solutions.u)
        if !((xₛₒₗ > SOL_MAX) || (yₛₒₗ > SOL_MAX) || (zₛₒₗ > SOL_MAX))
            if compare_results(solutions, parameters_dict_temp, equilibrium_type)
                push!(solutions_x, xₛₒₗ)
                push!(solutions_y, yₛₒₗ)
                push!(solutions_z, zₛₒₗ)
            end
        end
    end
    println("X ∈ $([round(e, digits=ROUND) for e in extrema(solutions_x)])")
    println("Y ∈ $([round(e, digits=ROUND) for e in extrema(solutions_y)])")
    println("Z ∈ $([round(e, digits=ROUND) for e in extrema(solutions_z)])")
    return nothing
end;

function get_common_bounds(data)
	stable_lst = intersect(data[2][:, 1], data[5][:, 1], data[8][:, 1])
	unstable_lst = intersect(data[3][:, 1], data[6][:, 1], data[9][:, 1])
	full_lst = union(stable_lst, unstable_lst)
    println("Full Bounds: [$(findmin(full_lst)[1]), $(findmax(full_lst)[1])]")
    println("Stable Bounds: [$(findmin(stable_lst)[1]), $(findmax(stable_lst)[1])]")
    println("Unstable Bounds: [$(findmin(unstable_lst)[1]), $(findmax(unstable_lst)[1])]")
    return nothing
end;

end

File: code_python/utils/compute_data.py
"""
Author: Rayla Kurosaki

GitHub: https://github.com/rkp1503
"""

import numpy as np
from scipy.integrate import odeint

SOL_MAX = 10


def solve_model(model, variables_dict, parameters_dict, t_max):
    ts = np.linspace(0, t_max, 10 * t_max + 1)
    variables_vals = list(variables_dict.values())
    parameter_vals = tuple(parameters_dict.values())
    return ts, odeint(model, variables_vals, ts, args=parameter_vals)


def get_bifurcation_data(model, variables_dict, parameters, t_max,
                         bifurcation_parameter, parameter_bounds,
                         parameter_values, data_points=10_000, debug=False):
    parameters_dict = dict(zip(parameters, parameter_values))
    x_t, x_s, x_u = [], [], []
    y_t, y_s, y_u = [], [], []
    z_t, z_s, z_u = [], [], []
    prev_x, prev_y, prev_z = "", "", ""
    curr_x, curr_y, curr_z = "", "", ""
    a, b = parameter_bounds[0], parameter_bounds[1]
    bifurcation_values = np.linspace(a, b, data_points + 1)
    for bifurcation_value in bifurcation_values:
        parameters_dict[bifurcation_parameter] = bifurcation_value
        ts, sol = solve_model(model, variables_dict, parameters_dict, t_max)
        n = int(len(ts) - np.floor(len(ts) * 0.95))
        x_min, x_max = min(sol[:, 0][-n:]), max(sol[:, 0][-n:])
        y_min, y_max = min(sol[:, 1][-n:]), max(sol[:, 1][-n:])
        z_min, z_max = min(sol[:, 2][-n:]), max(sol[:, 2][-n:])
        if val_in_bounds(x_min, x_max, data_points):
            x_t.append([bifurcation_value, x_min])
            x_t.append([bifurcation_value, x_max])
            if abs(x_min - x_max) <= 1 / data_points:
                curr_x = "Stable"
                x_s.append([bifurcation_value, x_min])
                x_s.append([bifurcation_value, x_max])
                pass
            else:
                curr_x = "Unstable"
                x_u.append([bifurcation_value, x_min])
                x_u.append([bifurcation_value, x_max])
                pass
            pass
        if val_in_bounds(y_min, y_max, data_points):
            y_t.append([bifurcation_value, y_min])
            y_t.append([bifurcation_value, y_max])
            if abs(y_min - y_max) <= 1 / data_points:
                curr_y = "Stable"
                y_s.append([bifurcation_value, y_min])
                y_s.append([bifurcation_value, y_max])
                pass
            else:
                curr_y = "Unstable"
                y_u.append([bifurcation_value, y_min])
                y_u.append([bifurcation_value, y_max])
                pass
            pass
        if val_in_bounds(z_min, z_max, data_points):
            z_t.append([bifurcation_value, z_min])
            z_t.append([bifurcation_value, z_max])
            if abs(z_min - z_max) <= 1 / data_points:
                curr_z = "Stable"
                z_s.append([bifurcation_value, z_min])
                z_s.append([bifurcation_value, z_max])
                pass
            else:
                curr_z = "Unstable"
                z_u.append([bifurcation_value, z_min])
                z_u.append([bifurcation_value, z_max])
                pass
            pass
        if (curr_x == curr_y == curr_z) and (curr_x != prev_x):
            if debug:
                if prev_x == "":
                    print(f"Start -> {curr_x}: {bifurcation_value}")
                    pass
                else:
                    print(f"{prev_x} -> {curr_x}: {bifurcation_value}")
                    pass
                pass
            prev_x = curr_x
            prev_y = curr_y
            prev_z = curr_z
            pass
        pass
    x_t = reformat_data(x_t)
    x_s = reformat_data(x_s)
    x_u = reformat_data(x_u)
    y_t = reformat_data(y_t)
    y_s = reformat_data(y_s)
    y_u = reformat_data(y_u)
    z_t = reformat_data(z_t)
    z_s = reformat_data(z_s)
    z_u = reformat_data(z_u)
    return x_t, x_s, x_u, y_t, y_s, y_u, z_t, z_s, z_u


def val_in_bounds(val_min, val_max, data_points):
    cond1 = 1 / data_points <= val_min < SOL_MAX
    cond2 = 1 / data_points <= val_max < SOL_MAX
    return cond1 and cond2


def reformat_data(data):
    if not data:
        return []
    return np.transpose(np.array(data))


def approximate_bounds():
    return None


def get_common_bounds(data):
    x_s, x_u = data[1][:, 0], data[2][:, 0]
    y_s, y_u = data[4][:, 0], data[5][:, 0]
    z_s, z_u = data[7][:, 0], data[8][:, 0]
    lst_s = list(x_s.intersection(y_s).intersection(z_s))
    lst_u = list(x_u.intersection(y_u).intersection(z_u))
    lst_t = list(set(lst_s) | set(lst_u))
    print(f"Full Bounds: [{min(lst_t)}, {max(lst_t)}]")
    print(f"Stable Bounds: [{min(lst_s)}, {max(lst_s)}]")
    print(f"Untable Bounds: [{min(lst_u)}, {max(lst_u)}]")
    return None
